### PR TITLE
(PE-28209) WIP Rough outline for apply in bolt-server

### PR DESCRIPTION
**Pull request by donoghuc**
Currently just a rough idea of the general shape.

TODO: Plugin management, tests, packaging

### CODE REVIEW

**Comment by donoghuc** [comment]
This is nowhere close. Just want to have something to discuss with team. It currenly can apply a catalog that does not require any plugins
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bundle exec scripts/server_client.rb 'use-default' ssh://bolt:bolt@172.18.0.1:20022 -a apply
{
  "target": "172.18.0.1",
  "action": "apply",
  "object": null,
  "status": "success",
  "value": {
    "report": {
      "host": "25151669e75f.corp.puppetlabs.net",
      "time": "2020-03-30T23:59:11.791157917+00:00",
      "configuration_version": 1581636379,
      "transaction_uuid": null,
      "report_format": 10,
      "puppet_version": "6.14.0",
      "status": "changed",
      "transaction_completed": true,
      "noop": false,
      "noop_pending": false,
      "environment": null,
      "logs": [
        {
          "level": "notice",
          "message": "hello world",
          "source": "Puppet",
          "tags": [
            "notice"
          ],
          "time": "2020-03-30T23:59:11.832634892+00:00",
          "file": null,
          "line": null
        },
        {
          "level": "notice",
          "message": "defined 'message' as 'hello world'",
          "source": "/Stage[main]/Main/Notify[hello world]/message",
          "tags": [
            "notice",
            "notify",
            "class"
          ],
          "time": "2020-03-30T23:59:11.836195423+00:00",
          "file": null,
          "line": 1
        },
        {
          "level": "notice",
          "message": "Applied catalog in 0.02 seconds",
          "source": "Puppet",
          "tags": [
            "notice"
          ],
          "time": "2020-03-30T23:59:11.849799849+00:00",
          "file": null,
          "line": null
        }
      ],
      "metrics": {
        "resources": {
          "name": "resources",
          "label": "Resources",
          "values": [
            [
              "total",
              "Total",
              1
            ],
            [
              "skipped",
              "Skipped",
              0
            ],
            [
              "failed",
              "Failed",
              0
            ],
            [
              "failed_to_restart",
              "Failed to restart",
              0
            ],
            [
              "restarted",
              "Restarted",
              0
            ],
            [
              "changed",
              "Changed",
              1
            ],
            [
              "out_of_sync",
              "Out of sync",
              1
            ],
            [
              "scheduled",
              "Scheduled",
              0
            ],
            [
              "corrective_change",
              "Corrective change",
              0
            ]
          ]
        },
        "time": {
          "name": "time",
          "label": "Time",
          "values": [
            [
              "notify",
              "Notify",
              0.003783742
            ],
            [
              "config_retrieval",
              "Config retrieval",
              0
            ],
            [
              "transaction_evaluation",
              "Transaction evaluation",
              0.007956162000482436
            ],
            [
              "catalog_application",
              "Catalog application",
              0.02113494899822399
            ],
            [
              "total",
              "Total",
              0.049016193
            ]
          ]
        },
        "changes": {
          "name": "changes",
          "label": "Changes",
          "values": [
            [
              "total",
              "Total",
              1
            ]
          ]
        },
        "events": {
          "name": "events",
          "label": "Events",
          "values": [
            [
              "total",
              "Total",
              1
            ],
            [
              "failure",
              "Failure",
              0
            ],
            [
              "success",
              "Success",
              1
            ]
          ]
        }
      },
      "resource_statuses": {
        "Notify[hello world]": {
          "title": "hello world",
          "file": null,
          "line": 1,
          "resource": "Notify[hello world]",
          "resource_type": "Notify",
          "provider_used": null,
          "containment_path": [
            "Stage[main]",
            "Main",
            "Notify[hello world]"
          ],
          "evaluation_time": 0.003783742,
          "tags": [
            "notify",
            "class"
          ],
          "time": "2020-03-30T23:59:11.832455812+00:00",
          "failed": false,
          "failed_to_restart": false,
          "changed": true,
          "out_of_sync": true,
          "skipped": false,
          "change_count": 1,
          "out_of_sync_count": 1,
          "events": [
            {
              "audited": false,
              "property": "message",
              "previous_value": "absent",
              "desired_value": "hello world",
              "historical_value": null,
              "message": "defined 'message' as 'hello world'",
              "name": "message_changed",
              "status": "success",
              "time": "2020-03-30T23:59:11.832579745+00:00",
              "redacted": null,
              "corrective_change": false
            }
          ],
          "corrective_change": false
        }
      },
      "corrective_change": false,
      "catalog_uuid": "5a4372c6-253f-46df-be99-3c40c9922423",
      "cached_catalog_status": "not_used"
    },
    "_output": "changed: 1, failed: 0, unchanged: 0 skipped: 0, noop: 0"
  }
}
```

File: lib/bolt_server/transport_app.rb [hunk 1]
```diff
@@ -53,9 +55,21 @@ def initialize(config)
       # This is needed until the PAL is threadsafe.
       @pal_mutex = Mutex.new
 
+      @apply_task = construct_apply_task
       super(nil)
     end
 
+    def construct_apply_task
```
Comment (donoghuc) [inline lib/bolt_server/transport_app.rb]: Should this go in `libexec`?

Given the apply_helpers task are basically all the same code (initializing a isolated puppet env and unpacking plugin tarball) I think for bolt-server I would like to just have a single task and add logic to modify behavior.

File: lib/bolt_server/transport_app.rb [hunk 2]
```diff
@@ -182,6 +196,15 @@ def run_script(target, body)
       [@executor.run_script(target, file_location, body['arguments'])]
     end
 
+    def apply_catalog(target, body)
+      # TODO: sync plugins to env cahce, tar them up, base-64 encode and pass to task
+      body['apply_options'] ||= {}
+      body['action'] = 'apply'
+      results = @executor.run_task(target, @apply_task, body)
```
Comment (donoghuc) [inline lib/bolt_server/transport_app.rb]: I think we will only ever have a single target at a time?

File: libexec/bolt_server_apply.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      catalog.environment = env.name.to_s
+      catalog.environment_instance = env
+      Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(facts, catalog)
+      catalog = catalog.to_ral
+      configurer = Puppet::Configurer.new
+      configurer.run(catalog: catalog, report: report, pluginsync: false)
+    end
+    puts JSON.pretty_generate(report.to_data_hash)
+    exit_code = report.exit_status != 1
+  else
```
Comment (donoghuc) [inline libexec/bolt_server_apply.rb]: TODO: add logic for handling `gather facts` and `gather resources`

File: scripts/server_client.rb [hunk 4]
```diff
@@ -65,8 +49,62 @@ def run_task(task_name, target_name, params, base_uri: 'https://localhost:62658'
...
+    exit 1
+  end
 
-puts JSON.pretty_generate(client.run_task(task, target, params))
+  opts.on("-a", "--action ENUM", %i[run_task run_command apply], "Run action") do |action|
+    case action
+    when :run_task
+      params = ARGV[2] || '{}'
+      params = JSON.parse(params)
+      puts JSON.pretty_generate(client.run_task(object, target, params))
```
Comment (donoghuc) [inline scripts/server_client.rb]: Probably add the rest of the actions. Helpful for quickly spinning up debug env

File: lib/bolt_server/transport_app.rb (hunk 2 above)
Comment (McdonaldSeanp) [reply lib/bolt_server/transport_app.rb:203]: Yep, only ever one target in these requests

**Comment by donoghuc** [comment]
Closing until we figure out our plan for pluginsync.

**Pull request closed without merging**